lib/services/watcher.go

```diff
@@ -1666,3 +1667,162 @@ func (p *accessRequestCollector) processEventAndUpdateCurrent(ctx context.Contex
+}
+
+// NewSAMLIdPServiceProviderWatcher returns a new instance of SAMLIdPServiceProviderWatcher. The context here will be used to
+// exit early from the resource watcher if needed.
+func NewSAMLIdPServiceProviderWatcher(ctx context.Context, cfg SAMLIdPServiceProviderWatcherConfig) (*SAMLIdPServiceProviderWatcher, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	collector := &samlIDPServiceProviderCollector{
+		Log:             cfg.RWCfg.Log,
+		Cfg:             cfg,
+		initializationC: make(chan struct{}),
+	}
+	watcher, err := newResourceWatcher(ctx, collector, cfg.RWCfg)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &SAMLIdPServiceProviderWatcher{
+		ResourceWatcher: watcher,
+		Collector:       collector,
+	}, nil
+}
+
+// SAMLIdPServiceProviderWatcher is built on top of resourceWatcher to monitor SAML IdP service provider resources.
+type SAMLIdPServiceProviderWatcher struct {
+	ResourceWatcher *resourceWatcher
+	Collector       *samlIDPServiceProviderCollector
+}
+
+// samlIDPServiceProviderCollector accompanies resourceWatcher when monitoring SAML IdP service provider resources.
+type samlIDPServiceProviderCollector struct {
+	// Log is a logger.
+	Log logrus.FieldLogger
+	// SAMLIdPServiceProviderWatcherConfig is the watcher configuration.
+	Cfg SAMLIdPServiceProviderWatcherConfig
+	// mu guards "current"
+	mu sync.RWMutex
+	// current holds a map of the currently known SAML IdP service provider resources.
+	current map[string]types.SAMLIdPServiceProvider
+	// initializationC is used to check that the watcher has been initialized properly.
+	initializationC chan struct{}
+	once            sync.Once
+}
+
+// resourceKind specifies the resource kind to watch.
+func (p *samlIDPServiceProviderCollector) resourceKind() string {
+	return types.KindSAMLIdPServiceProvider
+}
+
+// initializationChan is used to check if the initial state sync has been completed.
+func (p *samlIDPServiceProviderCollector) initializationChan() <-chan struct{} {
+	return p.initializationC
+}
+
+// getResourcesAndUpdateCurrent refreshes the list of current resources.
+func (p *samlIDPServiceProviderCollector) getResourcesAndUpdateCurrent(ctx context.Context) error {
+	var samlIDPServiceProviders []types.SAMLIdPServiceProvider
+	nextToken := ""
+	for {
+		var sps []types.SAMLIdPServiceProvider
+		var err error
+		sps, nextToken, err = p.Cfg.spService.ListSAMLIdPServiceProviders(ctx, p.Cfg.PageSize, nextToken)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		samlIDPServiceProviders = append(samlIDPServiceProviders, sps...)
+		if nextToken == "" {
+			break
+		}
+	}
+
+	newCurrent := make(map[string]types.SAMLIdPServiceProvider, len(samlIDPServiceProviders))
+	for _, samlIDPServiceProvider := range samlIDPServiceProviders {
+		newCurrent[samlIDPServiceProvider.GetName()] = samlIDPServiceProvider
+	}
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.current = newCurrent
+	p.defineCollectorAsInitialized()
```

Comment:
```suggestion
	p.mu.Lock()
	p.current = newCurrent
	p.defineCollectorAsInitialized()
	p.mu.Unlock()
```

Unless there's some reason I'm not spotting?

Reply:
I had *thought* that this was causing some ordering issues in the tests, but I realized later that it was the sorting issue in the tests. I'll reduce this.